letta/server/server.py

```diff
@@ -1750,3 +1752,107 @@ def get_agent_block_by_label(self, user_id: str, agent_id: str, label: str) -> B
             if block.label == label:
                 return block
         return None
+
+    def run_tool(self, tool_id: str, tool_args: str, user_id: str) -> FunctionReturn:
+        """Run a tool using the sandbox and return the result"""
+
+        try:
+            tool_args_dict = json.loads(tool_args)
+        except json.JSONDecodeError:
+            raise ValueError("Invalid JSON string for tool_args")
+
+        # Get the tool by ID
+        user = self.user_manager.get_user_by_id(user_id=user_id)
+        tool = self.tool_manager.get_tool_by_id(tool_id=tool_id, actor=user)
+        if tool.name is None:
+            raise ValueError(f"Tool with id {tool_id} does not have a name")
+
+        # TODO eventually allow using agent state in tools
+        agent_state = None
+
+        try:
+            sandbox_run_result = ToolExecutionSandbox(tool.name, tool_args_dict, user_id).run(agent_state=agent_state)
+            if sandbox_run_result is None:
+                raise ValueError(f"Tool with id {tool_id} returned execution with None")
+            function_response = str(sandbox_run_result.func_return)
+
+            return FunctionReturn(
+                id="null",
+                function_call_id="null",
+                date=get_utc_time(),
+                status="success",
+                function_return=function_response,
+            )
+        except Exception as e:
+            # same as agent.py
+            from letta.constants import MAX_ERROR_MESSAGE_CHAR_LIMIT
+
+            error_msg = f"Error executing tool {tool.name}: {e}"
+            if len(error_msg) > MAX_ERROR_MESSAGE_CHAR_LIMIT:
+                error_msg = error_msg[:MAX_ERROR_MESSAGE_CHAR_LIMIT]
+
+            return FunctionReturn(
+                id="null",
```

Comment:
@sarahwooders @mattzh72 if we use `FunctionReturn` as the response model, we need to fill these fields (`id`, `function_call_id`) in with dummy fields.

However, we could also create a new response model that doesn't have these fields, but still has the `status` and `function_return` fields?

Reply:
I think its fine to just set them with dummy fields